### ISSUE TITLE
Move `/homeless-shelter` somewhere where root can't write

### ISSUE THREAD

**Issue by NinjaTrappeur** [NixOS/nix#8313]
During a build, Nix sets `$HOME` to `/homeless-shelter` to prevent builds from messing with the build user `$HOME`. This works fine in a sandboxed build where the build worker can't access `/homeless-shelter. Any access to $HOME will lead a build to fail.

Sadly, in some situations, like when running Nix without build sandboxing in a unprivileged contained, the build user can access `/homeless-shelter`. We end up in a situation where a Nix build creates `/homeless-shelter` and write some files in it. [The subsequent unrelated builds then fail](https://github.com/NixOS/nix/blob/master/src/libstore/build/local-derivation-goal.cc#L752) (rightfully so) until something manually removes `/homeless-shelter`.

This behavior is annoying, it arises during the pip wheel generation for instance. Pip will try to access $HOME and create a cache directory. If this somehow fails, it'll fallback to use the current directory (ie. build root) as cache. It means that in some builds, the ability of writing to $HOME changes the build outcome. The pip wheel generation outcome then depends on whether or not the build is run in a sandboxed mode.

I think that for reproducibility sake, we should make sure that regardless whether the build is sandboxed or not, it is impossible for the builder to write in $HOME.

**Describe the solution you'd like**

I see two options here:

1. We can find a directory we're sure root won't be able to write to. `/proc` and `/sys` comes directly in mind.
2. We create a read-only mount for $HOME.

1 is trivial to implement and would likely be my preferred approach. But I feel like it's hacky and potentially not very portable.

I'm up to implement the fix, but I want to be sure this is a good approach and it has fair chances to be merged.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by SuperSandro2000** [comment]
> We create a read-only mount for $HOME.

Doesn't this also require extra permissions?



> We can find a directory we're sure root won't be able to write to. `/proc` and `/sys` comes directly in mind.

or ``/dev/null``.

**Comment by NinjaTrappeur** [comment]
> Doesn't this also require extra permissions?

True. I was thinking about the use case where the un-sanboxed build is triggered by root, but indeed, could be another user. Disregard that idea.

**Comment by thufschmitt** [comment]
That's indeed very annoying.

I've no idea whether using something like `/dev/null` would be a good solution. It feels a bit dangerous (is root _really_ unable to write to it?).

Alternatively, we could point `$HOME` to a subfolder of the build root. But that would probably be a bigger change

**Comment by NinjaTrappeur** [comment]
> Alternatively, we could point $HOME to a subfolder of the build root. But that would probably be a bigger change

Doing this would mean that builders can create and write to `$HOME` by default. I don't think this is desirable. I'd be more keen to keep the current sandboxed approach where a build fails when trying to create `$HOME`. This behavior still can be changed on a per-derivation basis by pointing `$HOME` somewhere else in the builder.

------------------------------

:+1:  for the rest, I'm unsure as well, that's why I created an issue to discuss that instead of filing a PR.
